rust/src/wrapper/device/mod.rs

```
@@ -230,35 +263,159 @@ impl Device {
+        let random_address_cmd = LeSetAdvertisingSetRandomAddressBuilder {
+            advertising_handle,
+            random_address,
+        };
+        self.send_command(random_address_cmd.into(), true).await?;
+
+        // set adv data
+        let advertising_data_cmd = LeSetExtendedAdvertisingDataBuilder {
+            advertising_data: adv_data.into_bytes(),
+            advertising_handle,
+            fragment_preference: FragmentPreference::ControllerMayFragment,
+            operation: Operation::CompleteAdvertisement,
+        };
+        self.send_command(advertising_data_cmd.into(), true).await?;
+
+        // enable adv
+        let extended_advertising_enable_cmd = LeSetExtendedAdvertisingEnableBuilder {
+            enable: Enable::Enabled,
+            enabled_sets: vec![EnabledSet {
+                advertising_handle,
+                duration: 0,
+                max_extended_advertising_events: 0,
+            }],
+        };
+        self.send_command(extended_advertising_enable_cmd.into(), true)
+            .await?;
+
+        self.advertising_status = AdvertisingStatus::AdvertisingExtended;
+        Ok(())
     }
 
     /// Stop advertising.
     pub async fn stop_advertising(&mut self) -> PyResult<()> {
         Python::with_gil(|py| {
-            self.0
+            self.obj
                 .call_method0(py, intern!(py, "stop_advertising"))
                 .and_then(|coroutine| into_future(coroutine.as_ref(py)))
         })?
         .await
-        .map(|_| ())
+        .map(|_| ())?;
+
+        if self.advertising_status == AdvertisingStatus::AdvertisingLegacy {
+            self.advertising_status = AdvertisingStatus::NotAdvertising;
+        }
+        Ok(())
+    }
+
+    /// Stop advertising extended.
+    #[cfg(feature = "unstable_extended_adv")]
+    pub async fn stop_advertising_extended(&mut self) -> PyResult<()> {
+        if AdvertisingStatus::AdvertisingExtended != self.advertising_status {
+            return Ok(());
```

Should this be return Error?

---

Same as above.

The Python legacy stop_advertising no-ops on duplicate calls to stop advertising.

Do you think it would be better to return an error?

---

Same as above, makes sense to keep behavior the same as the python

---

I think reflecting what Python does is fine, but let's sprinkle in some docs on these methods explaining the state transitions and expected return values.